Allow external structs.

mohammadfawaz opened the pull request:
Specifically,

1. Introduce a PlaintextType::ExternalStruct, and allow referring to such a type in .aleo code by a Locator.

2. Allow casting to external struct types.

3. Let two struct types be considered the same type as long as they are structurally the same and have the same local names. This funky ruleset is effectively here for backwards compatibility. In more detail:

```
/// Equivalence of structs means they have the same local names (regardless of whether
/// they're local or external), and their members have the same names and equivalent
/// types in the same order, recursively.
///
/// Equivalence of arrays means they have the same length and their element types are
/// equivalent.
///
/// This definition of equivalence was chosen to balance these concerns:
///
/// 1. All programs from before the existence of external structs will continue to work -
/// thus it's necessary for a struct created from another program to be considered equivalent
/// to a local one with the same name and structure, as in practice that was the behavior.
/// 2. We don't want to allow a fork. Thus we do need to check names, not just structural
/// equivalence - otherwise we could get a program deployable to a node which is using
/// this check, but not deployable to a node running an earlier SnarkVM.
```

Also note this other usecase from a user:

```
currently we got in an issue where one mainnet deployed contract imports tokenRegistry, and the contract we are writing for A imports that contract and A contract. but, A contract and tokenRegistryContract have same struct names so we cannot compile that.If we cannot change struct name in A contract, we need to wait for snarkVM update, just to compile the contracts
```

Commit:
External structs

Co-authored-by: Michael Benfield <mike.benfield@gmail.com>

mohammadfawaz commented:
Rebase and squash https://github.com/ProvableHQ/snarkVM/pull/2798 and opened a new PR with Michael as the co-author of the single commit. This will make it easier to keep rebasing as we review this.

Commit:
Add a test that uses an external struct in a record

mohammadfawaz commented:
@vicsn I added a test that uses an external struct inside a record: https://github.com/ProvableHQ/snarkVM/pull/3045/commits/4fb2cd37cbe094aa6bbc425163bf7e7e08c6bad6

vicsn commented on synthesizer/src/vm/tests/test_v13.rs:
```diff
@@ -0,0 +1,105 @@
+    // Define the first program with a record.
+    let program_one = Program::from_str(
+        r"
+program test_one.aleo;
```
Can you also test that we can now reference the struct of a program deployed **before** the new consensus version - the easiest will be to use e.g. the hardcoded `credits.aleo/committee_state`

vicsn replied:
The above idea was mine, but another ChatGPT idea: test having `[external struct; N]` somewhere.

mohammadfawaz replied:
Added two more e2e tests that use the two patterns above.

vicsn commented on synthesizer/program/src/traits/stack_and_registers.rs:
```diff
@@ -150,6 +151,109 @@ pub trait StackTrait<N: Network> {
+///
+/// The stacks are passed because struct types need to access their stack to get their
+/// structure.
+pub fn types_equivalent<N: Network>(
```
ChatGPT pointed me to the useful idea that we can easily add a couple obvious positive and negative tests for this central function.

mohammadfawaz replied:
Added a series of unit tests in `synthesizer/program/tests/types/equivalence.rs`

Commit:
- Move to consensus version 13
- Add a few more tests

mohammadfawaz commented:
Review comments addressed in https://github.com/ProvableHQ/snarkVM/pull/3045/commits/3e0e03cc52cefe5cef6eaf577863b13b55fbb9b2

vicsn commented:
Please no more force pushing after reviews have started

vicsn commented on synthesizer/program/src/traits/stack_and_registers.rs:
```diff
@@ -242,11 +242,16 @@ pub fn types_equivalent<N: Network>(
             struct_compare(&*external_stack, struct_type0, stack1, struct_type1)
         }
         (Struct(id), ExternalStruct(loc)) => {
+            dbg!("HERE!");
```
Now we know how the sausage is made.

mohammadfawaz replied:
OH no... isn't `HERE` universal though? 😆

vicsn reviewed (approved):
LGTM, one test seems to fail. If you run with `REWRITE_EXPECTATIONS=1`, please still check the new output makes sense.

Commit:
Remove dbg statements and update expectations of failing test

Commit:
Merge branch 'staging' into struct-namespace-squashed

d0cd commented on console/program/src/data_types/plaintext_type/bytes.rs:
```diff
@@ -23,7 +23,8 @@ impl<N: Network> FromBytes for PlaintextType<N> {
             2 => Ok(Self::Array(ArrayType::read_le(&mut reader)?)),
-            3.. => Err(error(format!("Failed to deserialize annotation variant {variant}"))),
+            3 => Ok(Self::ExternalStruct(Locator::read_le(&mut reader)?)),
+            4.. => Err(error(format!("Failed to deserialize annotation variant {variant}"))),
```
Exists in current implementation, but it might be more clear to say
```suggestion
            4.. => Err(error(format!("Failed to deserialize plaintext type variant {variant}"))),
```

d0cd commented on console/program/src/data_types/plaintext_type/mod.rs:
```diff
@@ -18,7 +18,7 @@ mod parse;
+use crate::{ArrayType, Identifier, LiteralType, Locator, ProgramID, StructType};
 use snarkvm_console_network::prelude::*;
 
 /// A `PlaintextType` defines the type parameter for a literal, struct, or array.
```
```suggestion
/// A `PlaintextType` defines the type parameter for a literal, struct, array, or external struct.
```

d0cd commented on console/program/src/data_types/plaintext_type/size_in_bits.rs:
```diff
@@ -67,7 +74,32 @@ impl<N: Network> PlaintextType<N> {
+                // Look up the struct
+                let struct_ = get_external_struct(locator)?;
+
+                // Account for the plaintext variant bits.
```
Could consider unifying with the struct logic in the variant above, for consistency. 
Ditto for `size_in_bits_raw`.

d0cd commented on synthesizer/process/src/stack/mod.rs:
```diff
@@ -332,6 +332,12 @@ impl<N: Network> Stack<N> {
             }
         }
 
+        for mapping in self.program.mappings().values() {
```
This check is performed in `verify.rs` gated by a consensus version. Does it also need to happen here?

d0cd commented on synthesizer/program/src/constructor/mod.rs:
```diff
@@ -51,6 +51,16 @@ impl<N: Network> ConstructorCore<N> {
 
+    /// Returns whether this constructor refers to an external struct.
+    pub fn contains_external_struct(&self) -> bool {
+        self.commands.iter().any(|command| {
```
Should we also check the other command variants?

d0cd commented on synthesizer/program/src/finalize/mod.rs:
```diff
@@ -80,6 +80,12 @@ impl<N: Network> FinalizeCore<N> {
+    pub fn contains_external_struct(&self) -> bool {
+        self.commands
+            .iter()
+            .any(|command| matches!(command, Command::Instruction(inst) if inst.contains_external_struct()))
```
Should we also check the other command variants?

d0cd commented on synthesizer/program/src/logic/instruction/mod.rs:
```diff
@@ -589,6 +589,16 @@ impl<N: Network> Instruction<N> {
+        // Only cast instructions may contain an explicit reference to an external struct.
+        // Calls may produce them, but they don't explicitly reference the type, and that's
+        // always been allowed.
+        matches!(self,
```
When a new instruction is added, we may forget to update this code.
For example, dynamic dispatch would in theory support `get.dynamic.record r0.microcredits into r1 as foo.aleo/data`, but it's easy for us to forget to update this rule.
This also holds true for some of the other checks that match on the instructions?
Is there a more reliable way we can enforce these properties?

d0cd commented on synthesizer/program/src/logic/instruction/operation/deserialize.rs:
```diff
@@ -535,6 +546,7 @@ where
             // Cache the plaintext bits, and return the struct.
             Ok(circuit::Plaintext::Struct(members, Default::default()))
         }
+        PlaintextType::ExternalStruct(_identifier) => todo!(),
```
Bailing with an error message is fine if this is not supported.

d0cd commented on synthesizer/program/src/logic/instruction/operation/deserialize.rs:
```diff
@@ -349,6 +355,7 @@ where
             // Cache the plaintext bits, and return the struct.
             Ok(Plaintext::Struct(members, Default::default()))
         }
+        PlaintextType::ExternalStruct(_identifier) => todo!(),
```
Bailing with an error message is fine if this is not supported.

d0cd commented on synthesizer/program/tests/types/equivalence.rs:
```diff
@@ -0,0 +1,236 @@
+    Ok(stack)
+}
+
+// ---------- 1️⃣ Literal equivalence ----------
```
Nit. We should get rid of the numerical emojis in this file, to support our friends whose editors can't render :)

d0cd commented on synthesizer/src/vm/tests/test_v13.rs:
```diff
@@ -0,0 +1,105 @@
+}
+
+// This test verifies that a program with external structs can be deployed on
+// consensus version 12.
```
```suggestion
// consensus version 13.
```

d0cd commented on synthesizer/src/vm/tests/test_v13.rs:
```diff
@@ -0,0 +1,105 @@
+    let caller_private_key = crate::vm::test_helpers::sample_genesis_private_key(rng);
+
+    // Initialize the VM at the correct height.
+    let v10_height = CurrentNetwork::CONSENSUS_HEIGHT(consensus_version).unwrap();
```
```suggestion
    let height = CurrentNetwork::CONSENSUS_HEIGHT(consensus_version).unwrap();
```

d0cd commented on synthesizer/src/vm/tests/test_v13.rs:
```diff
@@ -0,0 +1,105 @@
+
+    // Initialize the VM at the correct height.
+    let v10_height = CurrentNetwork::CONSENSUS_HEIGHT(consensus_version).unwrap();
+    let vm = crate::vm::test_helpers::sample_vm_at_height(v10_height, rng);
```
```suggestion
    let vm = crate::vm::test_helpers::sample_vm_at_height(height, rng);
```

d0cd commented on synthesizer/src/vm/verify.rs:
```diff
@@ -255,6 +256,21 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
+                if consensus_version < ConsensusVersion::V13 {
+                    ensure!(
+                        !deployment.program().contains_external_struct(),
+                        "Invalid deployment transaction '{id}' - external structs may only be used beginning with Consensus version 10"
```
```suggestion
                        "Invalid deployment transaction '{id}' - external structs may only be used beginning with `ConsensusVersion::V13`"
```

d0cd commented on synthesizer/tests/tests/vm/execute_and_finalize/external_struct_in_array.aleo:
```diff
@@ -0,0 +1,45 @@
+
+    // Build two identical committee_state structs
+    cast r0 r1 into r2 as credits.aleo/committee_state;
+    cast r0 r1 into r3 as credits.aleo/committee_state;
```
Do you think it would be worthwhile to test two arrays one with external and one with local?

d0cd reviewed (commented):
It's getting close, I'm aligned with this notion of partial nominal and structural equivalence. Left a few suggestions.

A couple points/questions:
- What is the import rule? If a program A uses an external struct in B which itself uses an external struct in C, does A have to import B and C? It would be great to document and test the rule.
- Before merging, we should refresh the program regressions and run it to make sure that the VM can still be initialized with all of the new checks.

Commit:
- Address more review comments
- Fix a bug related to type equivalence for nested external structs

mohammadfawaz commented:
@d0cd I believe I addressed all your comments in https://github.com/ProvableHQ/snarkVM/pull/3045/commits/93e4e4807f6e2c04fa1d2e93f6c1f42f352dc28d except for 

> This check is performed in verify.rs gated by a consensus version. Does it also need to happen here?

I'm not sure why the check is gated by a consensus version check in `verify.rs`. Do you remember why Michael added this? Also, if we do want to added a consensus version check in `synthesizer/process/src/stack/mod.rs` too, how do we go about that? How do we get the consensus version in that file?

> What is the import rule? If a program A uses an external struct in B which itself uses an external struct in C, does A have to import B and C? It would be great to document and test the rule.

I added a test that showcases that transitive dependencies are not required at the top level as long as the top level does not explicitly use the external struct name directly. I think this rule matches how `import` is meant to work for all symbols.

> Before merging, we should refresh the program regressions and run it to make sure that the VM can still be initialized with all of the new checks.

What does this mean exactly?

vicsn commented:
> I'm not sure why the check is gated by a consensus version check in verify.rs. Do you remember why Michael added this? Also, if we do want to added a consensus version check in synthesizer/process/src/stack/mod.rs too, how do we go about that? How do we get the consensus version in that file?

The consensus version check in verify.rs was added to ensure all validators agree external structs are only enabled from the new consensus version onwards. I think this comment's point is that the logic in synthesizer/process/src/stack/mod.rs might be duplicate and superfluous.

d0cd commented:
> > Before merging, we should refresh the program regressions and run it to make sure that the VM can still be initialized with all of the new checks.
> 
> What does this mean exactly?

The regression repo is [here](github.com/ProvableHQ/aleo-program-regressions).
It contains a few utilities we can run to make sure changes are backwards compatible.

Commit:
Merge branch 'staging' into struct-namespace-squashed

Commit:
Remove redundant check in stack/mod.rs

Commit:
Move testing of mapping with non-existent structs to `test_v13.rs` which is where we test consensus stuff

vicsn commented:
Thank you everyone :)
